src/game/AI/ScriptDevAI/scripts/northrend/storm_peaks.cpp

```diff
@@ -223,6 +223,141 @@ bool QuestAccept_npc_injured_miner(Player* pPlayer, Creature* pCreature, const Q
+            DisableTimer(FROSTWOLF_RESUME_SEARCH);
+            MoveToNewPoint(FROSTWOLF_MOVE_FINAL);
+        });
+        AddCustomAction(FROSTWOLF_CHANGE_DIRECTION, 15s, [&]()
+        {
+            m_creature->GetMotionMaster()->Clear();
+            m_creature->GetMotionMaster()->MoveIdle();
+            ResetTimer(FROSTWOLF_RESUME_SEARCH, 3s);
+            ResetTimer(FROSTWOLF_CHANGE_DIRECTION, 15s);
+        });
+        AddCustomAction(FROSTWOLF_RESUME_SEARCH, 0s, [&]()
+        {
+            switch (m_emoteCounter)
+            {
+                case 0:
+                case 1:
+                case 2:
+                {
+                    if (!m_creature->IsInCombat())
+                        DoBroadcastText(ethereal_frostworg_emotes[m_emoteCounter][urand(0,1)], m_creature);
+                    break;
+                }
+                case 3: break;
+                default: m_creature->ForcedDespawn(); break;
+            }
+            ++m_emoteCounter;
+            
+            MoveToNewPoint(FROSTWOLF_MOVE_NORMAL);
+        });
+    }
+
+    uint32 m_emoteCounter = 0;
+    bool m_summoned = false;
+
+    void MoveToNewPoint(uint32 type)
+    {
+        float dist = type == 1 ? 50.f : 20.f;
+        float x, y, z = 0.f;
+        while (abs(z - m_creature->GetPositionZ()) > 5)
+            m_creature->GetRandomPoint(m_creature->GetPositionX(), m_creature->GetPositionY(), m_creature->GetPositionZ(), dist, x, y, z, 10.f);
+        m_creature->GetMotionMaster()->MovePoint(type, Position(x, y, z), FORCED_MOVEMENT_RUN);
+    }
+
+    void Reset() override
+    {
+        MoveToNewPoint(FROSTWOLF_MOVE_NORMAL);
+    }
+
+    void SummonedCreatureJustDied(Creature* summoned) override
+    {
+        if (summoned->GetEntry() == NPC_STORMFORGED_INFILTRATOR)
+            m_creature->ForcedDespawn();
+    }
+
+    void JustSummoned(Creature* summoned) override
+    {
+        switch (urand(0,2))
+        {
+            case 0: DoBroadcastText(30916, summoned); break;
+            case 1: DoBroadcastText(30917, summoned); break;
+            case 2: DoBroadcastText(30918, summoned); break;
+        }
+    }
+
+    void MovementInform(uint32 movementType, uint32 uiPointId) override
+    {
+        switch (uiPointId)
+        {
+            case FROSTWOLF_MOVE_NORMAL:
+            {
+                MoveToNewPoint(FROSTWOLF_MOVE_NORMAL);
+                break;
+            }
+            case FROSTWOLF_MOVE_FINAL:
+            {
+                if (m_summoned)
+                    return;
+                m_summoned = true;
+                m_creature->GetMotionMaster()->Clear(true, true);
+                AddCustomAction(FROSTWOLF_SUMMON_INFILTRATOR, 2s, [&]() { DoCastSpellIfCan(nullptr, SPELL_SUMMON_INFILTRATOR); });
+                DoBroadcastText(ethereal_frostworg_emotes[3][0], m_creature);
+                break;
+            }
+            default: break;
+        }
+    }
+
+    void UpdateAI(const uint32 diff) override
```

Comment:
Make it so it doesnt need to override UpdateAI at all. Otherwise spell lists and other intelligent features wont be addable down the road if needed.

Reply:
I have no idea how

Reply:
you yourself have personally told me that I must not under any circumstances edit running timers and since there is no other way of pausing timers than to just not update these timers by intercepting the updates in UpdateAI that's all I can do.

Reply:
Except... I guess I could create a CustomAction that runs every 500ms and calls a function... let's call it... PupdateAI() while out of combat with a value that expresses how much time has passed, lets call it doff, which in turn then counts internal counters and keeps comparing them to timers of things I think I'll call CostumeActions and pops them off the queue once the counter has reached sufficient height...

Reply:
And have you thought about adding OOC - only custom actions? :D

Reply:
I have, but for what? It's one creature that needs them. To add a system like that for a single creature seems wasteful

Reply:
We already have combat-only actions, ooc only are not that big of a deal.

Reply:
If there are use cases for that, sure, but if there aren't, why?